compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1054,22 +1054,107 @@ private boolean validateSpecifiedFields(BLangRecordLiteral mappingConstructor, B
         return isFieldsValid;
     }
 
-    private boolean validateRequiredFields(BRecordType type, List<RecordLiteralNode.RecordField> specifiedFields,
-                                           DiagnosticPos pos) {
+    private boolean validateRequiredAndReadonlyFields(BRecordType type,
+                                                      List<RecordLiteralNode.RecordField> specifiedFields,
+                                                      DiagnosticPos pos) {
         HashSet<String> specFieldNames = getFieldNames(specifiedFields);
         boolean hasAllRequiredFields = true;
+        boolean hasValidReadonlyFields = true;
 
         for (BField field : type.fields) {
-            // Check if `field` is explicitly assigned a value in the record literal
-            // If a required field is missing, it's a compile error
-            if (!specFieldNames.contains(field.name.value) && Symbols.isFlagOn(field.symbol.flags, Flags.REQUIRED)) {
+            String fieldName = field.name.value;
+            BType fieldType = symTable.semanticError;
+            DiagnosticPos fieldPos;
+
+            if (specFieldNames.contains(fieldName)) {
+                RecordLiteralNode.RecordField recordField = getRecordFieldByName(specifiedFields, fieldName);
+                if (recordField == null) {
```

Comment:
Can this be `null` here, if the mapping constructor contains the field?

Reply:
Yeah not in this context, I've removed the null check.